src/main/java/com/lithium/dbi/rdbi/recipes/scheduler/DedupJobSchedulerDAO.java

```diff
@@ -0,0 +1,141 @@
+        "if isPaused then\n" +
+        "    return reserved\n" +
+        "end\n" +
+        "local reservedIndex = 1\n" +
+        "local nextLimit = tonumber($limit$)\n" +
+        "local nextOffset = 0\n" +
+        "while nextLimit > 0 do\n" +
+        "   local jobs = redis.call('ZRANGEBYSCORE', $readyQueue$, 0, $now$, 'WITHSCORES', 'LIMIT', nextOffset, nextLimit)\n" +
+        "   if next(jobs) == nil then\n" +
+        "       return reserved\n" +
+        "   end\n" +
+        "   for i=1,2*#jobs,2 do\n" +
+        "      local inRunningQueue = redis.call('ZSCORE', $runningQueue$, jobs[i])\n" +
+        "      if not inRunningQueue then\n" +
+        "          reserved[reservedIndex] = jobs[i]\n" +
+        "          reserved[reservedIndex + 1] = jobs[i + 1]\n" +
+        "          reservedIndex = reservedIndex + 2\n" +
+        "          redis.call('ZREM', $readyQueue$, jobs[i])\n" +
+        "          redis.call('ZADD', $runningQueue$, $ttl$, jobs[i])\n" +
+        "          nextLimit = nextLimit - 1\n" +
+        "      end\n" +
+        "   end\n" +
+        "   nextOffset = nextOffset + nextLimit\n" +
+        "end\n" +
+        "return reserved"
+    )
+    List<TimeJobInfo> reserveJobs(
+            @BindKey("readyQueue") String readyQueue,
+            @BindKey("runningQueue") String runningQueue,
+            @BindKey("pausedTube") String pausedTube,
+            @BindArg("limit") int limit,
+            @BindArg("now") long now,
+            @BindArg("ttl") long ttl);
+
+    /**
+     * Jobs in the ready queue are scored/sorted by when they should run. A ready job
+     * is considered expired if its score/scheduled run time is less than equal to the
+     * expirationTimeInMillis. As a result expirationTimeInMillis should be set to
+     * now - (the threshold past which a job sitting waiting to be reserved is considered expired).
+     *
+     * Jobs in the running queue are sorted by their time to live (ttl). A running job
+     * is considered expired if the current time is past the ttl. So expirationTimeInMillis
+     * should be set to now.
+     */
+    @Mapper(TimeJobInfoListMapper.class)
+    @Query(
+        "local expiredJobs = redis.call('ZRANGEBYSCORE', $queue$, 0, $expirationTimeInMillis$, 'WITHSCORES')\n" +
+        "redis.call('ZREMRANGEBYSCORE', $queue$, 0, $expirationTimeInMillis$)\n" +
+        "return expiredJobs"
+    )
+    List<TimeJobInfo> removeExpiredJobs(@BindKey("queue") String queue,
+                                        @BindArg("expirationTimeInMillis") Long expirationTimeInMillis);
+
+    /**
+     * Deleting a job deletes it from the scheduler in its entirety.
+     */
+    @Query(
+        "local deletedFromReadyQueue = redis.call('ZREM', $readyQueue$, $job$)\n" +
+        "local deletedFromRunningQueue = redis.call('ZREM', $runningQueue$, $job$)\n" +
+        "if deletedFromReadyQueue == 0 and deletedFromRunningQueue == 0 then\n" +
```

Comment:
this is wrong. We only want to delete from the running queue, right ?  Wouldn't this also delete the pending ready queue job, if one existed ?